compiler/hie-core/src/Development/IDE/LSP/LanguageServer.hs

```diff
@@ -100,9 +127,17 @@ runLanguageServer options userHandlers getIdeState = do
                                 "Message: " ++ show x ++ "\n" ++
                                 "Exception: " ++ show e
                     Response x@RequestMessage{_id, _params} wrap act ->
+                        flip finally (clearReqId _id) $
                         catch (do
-                            res <- act lspFuncs ide _params
-                            sendFunc $ wrap $ ResponseMessage "2.0" (responseId _id) (Just res) Nothing
+                            cancelOrRes <- race (waitForCancel _id) $ act lspFuncs ide _params
```

Comment:
I'd be tempted to use the optimisation of doing a quick "is it in the cancel set" before calling `act`. I'm worry that doing this way encourages a lot of async exceptions, and we know that certain places (e.g. TH) go wrong with such exceptions. That said, this approach will quickly weed out such issues, so maybe that's a good thing.

Reply:
I considered doing that but then decided against it exactly for the reasons you mentioned. I’d rather see things blow up and then fix them properly than try to hide them.

Reply:
Perhaps write a comment to that effect?

Reply:
Good idea, added a comment.